src/main/java/net/testudobank/MvcController.java

```diff
@@ -619,23 +612,167 @@ public String submitTransfer(@ModelAttribute("user") User sender) {
+    String withdrawResponse = submitWithdraw(user);
+
+    if (withdrawResponse.equals("account_info")) {
+
+      // create an entry for Ethereum if necessary
+      if (!TestudoBankRepository.getCustomerCryptoBalance(jdbcTemplate, userID, CRYPTO_NAME).isPresent()) {
+        TestudoBankRepository.initCustomerCryptoBalance(jdbcTemplate, userID, CRYPTO_NAME);
+      }
+
+      TestudoBankRepository.increaseCustomerCryptoBalance(jdbcTemplate, userID, CRYPTO_NAME, cryptoAmountToBuy);
+      TestudoBankRepository.insertRowToCryptoLogsTable(jdbcTemplate, userID, CRYPTO_NAME, CRYPTO_HISTORY_BUY_ACTION, currentTime, cryptoAmountToBuy);
+
+      updateAccountInfo(user);
+
+      return "account_info";
+    } else {
+      return "welcome";
+    }
+
   }
 
   /**
-   * 
+   * HTML POST request handler for the Sell Crypto Form page.
+   * <p>
+   * The same username+password handling from the login page is used.
+   * <p>
+   * If the password attempt is correct, and the purchase amount is a valid amount
+   * that does not exceed crypto balance, the cost of the cryptocurrency in cash will be
+   * added to the users cash balance, and cryptocurrency will be subtracted from the users account
+   * <p>
+   * If the password attempt is incorrect or the amount to purchase is invalid,
+   * the user is redirected to the "welcome" page.
+   * <p>
+   * Crypto purchase function is implemented by re-using deposit handler.
+   * Logic of deposit (applying to overdraft, adding to balance, etc.) is delegated to this handler.
+   *
    * @param user
    * @return "account_info" page if sell successful. Otherwise, redirect to "welcome" page.
    */
   @PostMapping("/sellcrypto")
   public String sellCrypto(@ModelAttribute("user") User user) {
-    return "welcome";
+
+
+    String userID = user.getUsername();
+    String userPasswordAttempt = user.getPassword();
+    String userPassword = TestudoBankRepository.getCustomerPassword(jdbcTemplate, userID);
+
+    //// Invalid Input/State Handling ////
+
+    // unsuccessful login
+    if (!userPasswordAttempt.equals(userPassword)) {
+      return "welcome";
+    }
+
+    double cryptoAmountToSell = user.getAmountToSellCrypto();
+    if (cryptoAmountToSell <= 0) {
+      return "welcome";
+    }
+
+    Optional<Double> cryptoBalance = TestudoBankRepository.getCustomerCryptoBalance(jdbcTemplate, userID, CRYPTO_NAME);
+
+    // possible for user to not have any crypto
+    if (!cryptoBalance.isPresent()) {
+      return "welcome";
+    }
+
+    // check if user has required crypto balance
+    // TODO: comparing doubles like this is probably not a good idea
+    if (cryptoBalance.get() < cryptoAmountToSell) {
+      return "welcome";
+    }
+
+    double cashValueOfCrypto = cryptoPriceClient.getCurrentEthValue() * cryptoAmountToSell;
+
+    String currentTime = SQL_DATETIME_FORMATTER.format(new java.util.Date());
```

Comment:
Nitpick: Probably a bit better for this to go after the 2 `user.set()` lines below and directly before `submitDeposit()`. Not a big deal.